data_structures/binary_tree/wavelet_tree.py

```diff
@@ -0,0 +1,207 @@
+from typing import List, Optional
+
+"""
+Wavelet tree is a data-structure designed to efficiently answer various range queries
+for arrays. Wavelets trees are different from other binary trees in the sense that
+the nodes are split based on the actual values of the elements and not on indices,
+such as the with segment trees or fenwick trees. You can read more about them here:
+1. https://users.dcc.uchile.cl/~jperez/papers/ioiconf16.pdf
+2. https://www.youtube.com/watch?v=4aSv9PcecDw&t=811s
+3. https://www.youtube.com/watch?v=CybAgVF-MMc&t=1178s
+"""
+
+
+class Node:
+    def __init__(self, length: int) -> None:
+        self.minn: int = -1
+        self.maxx: int = -1
+        self.map_left: List = [-1] * length
+        self.left: Optional[Node] = None
+        self.right: Optional[Node] = None
+
+    def __repr__(self) -> str:
+        return f"min_value: {self.minn}, max_value: {self.maxx}"
```

Comment:
Let's add a few doctests to placate @algorithms-keeper
```suggestion
        """
        >>> node = Node(length=27)
        >>> repr(node)
        'min_value: -1, max_value: -1'
        >>> repr(node) == str(node)
        True
        """
        return f"min_value: {self.minn}, max_value: {self.maxx}"
```